op-batcher/batcher/driver.go

```diff
@@ -264,49 +262,54 @@ func (l *BatchSubmitter) calculateL2BlockRangeToStore(ctx context.Context) (eth.
 func (l *BatchSubmitter) loop() {
 	defer l.wg.Done()
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	ticker := time.NewTicker(l.PollInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
 			l.loadBlocksIntoState(l.ctx)
+			l.publishStateToL1(ctx)
+		case <-l.done:
+			return
+		}
```

Comment:
Because of random `select` channel selection, we need to give priority to `l.done`. Also, don't we still want to publish the state once more after closing `l.done`? Otherwise it might happen that a stop request comes in and we just return here instead of publishing remaining frames.
```suggestion
		select {
		case <-ticker.C:
			select {
		        case <-l.done:
				l.publishStateToL1(ctx)
				return
			default:
			}
			
			l.loadBlocksIntoState(l.ctx)
			l.publishStateToL1(ctx)
		case <-l.done:
			l.publishStateToL1(ctx)
			return
		}
```

Reply:
Should we instead move the close check from the bottom of the `publishStateToL1` function to the top? I think that'd take care of closing it at the right time. With that being true (& synchronous tx submission), I don't think we'd actually get a case where we would close a pending channel.

Reply:
Interesting point about the random `select` ordering. There's a few assumptions in the current implementation:
1. `l.ctx` is cancelled when `l.done` is readable (closed). This is because `Stop` calls `l.cancel()` before `close(l.done)`
2. A cancelled `l.ctx` context will effectively prevent `l.loadBlocksIntoState(l.ctx)` from having any meaningful side effects.
3. `l.publishStateToL1(ctx)` will also have no effect without any loaded blocks.

This still could result in at least one unnecessary cycle, potentially multiple if these methods take longer than expected or the `ticker` speed is sufficiently fast. I can reorder this code to prioritize the `done` channel :+1: